### DISCUSSION TITLE
Video not loading on iOS

### DISCUSSION THREAD

**Discussion by adamo** [swup/swup#838] [Help] [answered]
### Describe your problem 🧐

Hi. When switching between pages the `html` `video` isn't loading.  As far as I can tell it only happens on iOS devices (tested in chrome, safari). Has anyone had this issue?

### What is your question? 🙋‍♀️

Is there a workaround for this? I've tried using `.mov` instead of mp4 but that didn't change anything.

### URL to reduced test case 🔗

https://cedo.depcore.agency

### Checked all these? 📚

- [X] I have read and searched the [official docs](https://swup.js.org)
- [X] I have checked [discussions](https://github.com/orgs/swup/discussions) and [existing issues](https://github.com/issues?q=is:issue+org:swup) for answers to my question
- [X] I agree to follow this project's [Code of Conduct](https://github.com/swup/.github/blob/master/CODE_OF_CONDUCT.md)
- [X] I have provided all necessary information to the best of my knowledge

**Comment by hirasso** [comment]
Hi @adamo , your problem is very likely not related to swup. A thing that always makes sense in these cases is to try to get it to work *without swup* first. In your case, you could try to add the [`playsinline`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video#playsinline) attribute to the video:

```html
<video autoplay muted loop playsinline poster="https://cedo.depcore.agency/themes/cedo//assets/images/FFFFFF-0.png">
    <source src="https://cedo.depcore.agency/themes/cedo//assets/videos/1.mp4" type="video/mp4">
</video>
```

**Reply by adamo** [reply]
I've added this previously but it didn't work the video wasn't loading either way. That's why i thought it could be a swup related problem/

**Reply by hirasso** [reply]
Thanks for the clarification!

**Comment by daun** [comment]
Does it also fail to load on initial load or only after page transitions triggered by swup?

**Reply by adamo** [reply]
Only after transition

**Comment by hirasso** [comment]
Safari has a history of blocking autoplaying video quite aggressively. Blame it on the advertising industry ;)

From the [docs](https://developer.apple.com/documentation/webkit/delivering_video_content_for_safari/):

> By default, autoplay executes only if the video **doesn’t contain an audio track**, or if the video element includes the muted attribute. Video playback pauses if the element gains an audio track, becomes unmuted without user interaction, **or if the video is no longer onscreen (either by CSS or due to the user scrolling the page)**. [...]

Following this, I'd try three things. 

1. Using [handbrake](https://handbrake.fr/) or a similar tool, try **completely removing the audio track** from the video. I can vaguely remember a case some time ago where only the `muted` attribute didn't work, I actually had to get rid of the audio track in the actual file.
2. If this doesn't fix it, you could try to start the video using JavaScript as soon as it's fully faded in (not tested myself):

```js
/** runs each time swup has fully rendered and faded in a page */
swup.hooks.on('visit:end', () => {
  document.querySelectorAll('video[autoplay]').forEach(el => el.play());
});
```
3. If this still doesn't work, my final idea would be to try to programmatically start the video using an [IntersectionObserver](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API).

Safari on iOS has been changing their handling of videos quite a few times, so don't expect this to continue to work next year, should you get it to work now...

**Reply by adamo** [reply]
The videos on first run load normally and are autoplayed (the muted attribute is added). Just after usung swup they don't load.

**Comment by hirasso** [comment]
@adamo I went ahead and created a [reduced test case](https://swup-demo-video-autoplay--hirasso.repl.co/) using your video. Turns out it had something to do with your video file. Using your original file, autoplay didn't work for me after a swup page visit, either (click on "Next Page", click on "Back"). I used [handbrake](https://handbrake.fr/) to convert the file: 

- removed the audio track:

![Screenshot 2023-11-26 at 15 54 41](https://github.com/swup/swup/assets/869813/63ba9a93-972c-45e6-9692-83611eb95f53)

- ticked "Web optimized":

![Screenshot 2023-11-26 at 15 54 02](https://github.com/swup/swup/assets/869813/ae3bdeeb-c6e2-4e69-a8bd-452a9f1e95b3)

This is the converted video file:

https://swup-demo-video-autoplay--hirasso.repl.co/video-1-handbrake.mp4

After I switched out the newly encoded video, it's now working for me. I'm afraid I don't know what exactly fixed it, but it definitely has something to do with a combination of these:

1. Safari on iOS
2. DOM updated by JavaScript
3. The encoding of the video file

Reduced test cases really are the gold standard for debugging issues like these. We'll need to advertise them more in the future.

**Reply by adamo** [reply]
Thank you very much. I'll try this.

**Reply by adamo** [reply, marked as answer]
That didn't work for me so I've finally ended up doing what the scripts plugin does but for video and hooked into `page:view` this script
```
       const videos = document.querySelectorAll("video");
        videos.forEach((video) => {
            video.load();
        });
```

**Comment by hirasso** [comment]
Side note: chrome on iOS is *also* just safari due to bad behaviour by Apple ☹️

**Comment by daun** [comment]
One thing to double-check is that the device in questions isn't in power-saving mode when testing this behavior. I ran into this once and it took me a few hours to figure out why it wouldn't work on the unplugged MacBook but worked fine on the stationary iMac.

**Comment by mathieupreaud** [comment]
I had the same issue on mobile (tested with iOS), but also on computer (Mac). In my case, the issue didn't occur with Swup v3, it only happened with Swup v4.

**Reply by daun** [reply]
Interesting. Can you share a link to a demo site or reproduction somewhere? I'd like to look into this if it turns out to be a regression in v4 after all.

**Reply by mathieupreaud** [reply]
Unfortunately, right now I don't have any demo site to reproduce the issue.
